### PR TITLE
Limit conditions for inferencing a possible FuncDecl

### PR DESCRIPTION
**Pull request by mininny**
Follow-up from https://github.com/apple/swift-syntax/issues/2109

```
class MyClass {
  word1 // Parsed as FunctionDecl
}
```

Currently, the parser consumes the `word1` token as an incomplete function declaration since most declarations within a MemberDeclList are functions. 

However, without more information around the token like `func` or `()` which would indicate that it is a possible FuncDecl, it could also be other declarations, like `let`, `var`, or even something like `typealias`.

This PR peeks the tokens after the identifier and checks if it has `(` or `<` which would indicate that it's a function. 

Any freestanding identifiers as shown in the example above would be parsed as `MissingDeclSyntax`, which spits out `unexpected 'word1' in MyClass`.

This would affect existing test cases that rely on the previous behavior and change the resulting diagnostics.

---

I've validated that this change is compatible with most test cases, but I'm not sure how(if I'm allowed to) to fix existing test cases, and add test cases(because many of them are translations from swift repo)

Also, I think a better diagnostic in the example above is something like: ``expected keyword before `word1`. did you mean to add `func` or `var`?``, but I'm not yet sure how I can feed the parsed information from Parser to Diagnostics.

### CODE REVIEW

File: Sources/SwiftParser/Declarations.swift [hunk 1]
```diff
@@ -281,9 +281,11 @@ extension Parser {
         )
       }
 
-      let isProbablyFuncDecl = self.at(.identifier, .wildcard) || self.at(anyIn: Operator.self) != nil
+      let isPossibleFuncIdentifier = self.at(.identifier, .wildcard)
+      let isPossibleFuncParen = self.peek(isAt: .identifier, .leftParen) || self.peek(isAt: .identifier, .leftAngle)
```
Comment (ahoppen) [inline Sources/SwiftParser/Declarations.swift:285]: You should be able to shorten this to

```suggestion
      let isPossibleFuncParen = self.peek(isAt: .identifier, .leftParen, .leftAngle)
```

**Review by ahoppen — COMMENTED** [review]
Thanks for the PR @mininny. The changes look good to me and I quite like the `unexpected 'word1'` diagnostic because at this point you can just view `word1` as garbage text inside the member decl list.

And you can definitely change the existing tests. If the diagnostics improve, it’s totally fine to change their expected behavior.

File: Sources/SwiftParser/Declarations.swift (hunk 1 above)
Comment (mininny) [reply Sources/SwiftParser/Declarations.swift]: Thanks! I shortened it. However, while testing it out, I found that in
```
class MyClass { 
    foo<Int>
}
```
`<` is parsed as a binaryOperator rather than a left angle. I think this might need some follow-up item that should be looked at.

**Comment by mininny** [comment]
> Thanks for the PR @mininny. The changes look good to me and I quite like the `unexpected 'word1'` diagnostic because at this point you can just view `word1` as garbage text inside the member decl list.
> 
> And you can definitely change the existing tests. If the diagnostics improve, it’s totally fine to change their expected behavior.

Thanks for the review and comment! I've added some tests and marked the PR as ready for review. 

I have another question; when adding tests, is it okay to add tests to `SwiftParserTests/translated`? I see that it's translated from the original swift repo.

**Review by ahoppen — APPROVED** [review]
Thanks. LGTM

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
It looks like your changes aren’t formatted. Could you run `format.py`?

**Comment by mininny** [comment]
> It looks like your changes aren’t formatted. Could you run `format.py`?

Yep, I ran the the format.py :+1:

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows
